### PR TITLE
ASoC: SOF: loader: add firmware size check on loading

### LINKED ISSUE #428: Upstream: Patch 9/14: loader size/alignment issues
feedback from Takashi Iwai

````text
> +/* generic module parser for mmaped DSPs */
> +int snd_sof_parse_module_memcpy(struct snd_sof_dev *sdev,
> +				struct snd_sof_mod_hdr *module)
> +{
> +	struct snd_sof_blk_hdr *block;
> +	int count;
> +	u32 offset;
> +
> +	dev_dbg(sdev->dev, "new module size 0x%x blocks 0x%x type 0x%x\n",
> +		module->size, module->num_blocks, module->type);
> +
> +	block = (void *)module + sizeof(*module);
> +
> +	for (count = 0; count < module->num_blocks; count++) {

Need a sanity check that it won't go beyond the actual firmware size.
User may pass a malicious module data, e.g. with extra large
num_blocks.
````

````text
> +		switch (block->type) {
> +		case SOF_BLK_IMAGE:
> +		case SOF_BLK_CACHE:
> +		case SOF_BLK_REGS:
> +		case SOF_BLK_SIG:
> +		case SOF_BLK_ROM:
> +			continue;	/* not handled atm */
> +		case SOF_BLK_TEXT:
> +		case SOF_BLK_DATA:
> +			offset = block->offset;
> +			break;
> +		default:
> +			dev_err(sdev->dev, "error: bad type 0x%x for block 0x%x\n",
> +				block->type, count);
> +			return -EINVAL;
> +		}
> +
> +		dev_dbg(sdev->dev,
> +			"block %d type 0x%x size 0x%x ==>  offset 0x%x\n",
> +			count, block->type, block->size, offset);
> +
> +		snd_sof_dsp_block_write(sdev, offset,
> +					(void *)block + sizeof(*block),
> +					block->size);
> +
> +		/* next block */
> +		block = (void *)block + sizeof(*block) + block->size;

This may lead to an unaligned access.
Also how is the endianess guaranteed?
````

### PR DESCRIPTION
**Pull request by bardliao**
User may want to load an invalid firmware with incorrect size value
in the header. So we need to make sure we won't read data beyond the
firmware data.

Signed-off-by: Bard liao <yung-chuan.liao@linux.intel.com>

Fix #428

### CODE REVIEW

File: sound/soc/sof/loader.c [hunk 1]
```diff
@@ -184,13 +193,19 @@ static int load_modules(struct snd_sof_dev *sdev, const struct firmware *fw)
 
 	/* parse each module */
 	module = (void *)fw->data + sizeof(*header);
+	remaining = fw->size;
```
Comment (plbossart) [inline sound/soc/sof/loader.c]: Is this correct?
fw->size equals header->file_size + sizeof(*header)), and you skip the header by doing
module = (void *)fw->data + sizeof(*header);

so,  should it be
remaining = fw->size - sizeof(*header?

Comment (bardliao) [reply sound/soc/sof/loader.c]: Thanks @plbossart . Yes, it should be remaining = fw->size - sizeof(*header). I also did a slight modification for readability.

**Comment by bardliao** [comment]
@plbossart I think it won't be endless now. From ```load_modules()``` we guarantee that module->size < fw->size and in ```snd_sof_parse_module_memcpy()``` the remaining value will be decrease in each cycle. So it will exit the for loop when there is no fw data remaining.

**Comment by plbossart** [comment]
> @plbossart I think it won't be endless now. From `load_modules()` we guarantee that module->size < fw->size and in `snd_sof_parse_module_memcpy()` the remaining value will be decrease in each cycle. So it will exit the for loop when there is no fw data remaining.

Ah nice language misunderstanding. I was talking about Takashi's second comment 
````
> +		/* next block */
> +		block = (void *)block + sizeof(*block) + block->size;

This may lead to an unaligned access.
Also how is the endianess guaranteed?
````

Endianess don't refer to the 'end' but the order of the data within a block/access unit (Little or Big?)

**Pull request merged**